### PR TITLE
[SPARK-24391][SQL] Support arrays of any types by from_json

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

The PR removes a restriction for element types of array type which exists in `from_json` for the root type. Currently, the function can handle only arrays of structs. Even array of primitive types is disallowed. The PR allows arrays of any types currently supported by JSON datasource. Here is an example of an array of a primitive type:

```
scala> import org.apache.spark.sql.functions._
scala> val df = Seq("[1, 2, 3]").toDF("a")
scala> val schema = new ArrayType(IntegerType, false)
scala> val arr = df.select(from_json($"a", schema))
scala> arr.printSchema
root
 |-- jsontostructs(a): array (nullable = true)
 |    |-- element: integer (containsNull = true)
```
and result of converting of the json string to the `ArrayType`:
```
scala> arr.show
+----------------+
|jsontostructs(a)|
+----------------+
|       [1, 2, 3]|
+----------------+
```

## How was this patch tested?

I added a few positive and negative tests:
- array of primitive types
- array of arrays
- array of structs
- array of maps

### CODE REVIEW

**Comment by maropu** [comment]
better to add tests in `json-functions.sql`?

**Comment by maropu** [comment]
Can we also accept primitive arrays in `to_json`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 1]
```diff
@@ -523,6 +523,8 @@ case class JsonToStructs(
   // can generate incorrect files if values are missing in columns declared as non-nullable.
   val nullableSchema = if (forceNullableSchema) schema.asNullable else schema
 
+  val unpackArray: Boolean = options.get("unpackArray").map(_.toBoolean).getOrElse(false)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: private? (This is not related to this pr though, `nullableSchema` also can be private?)

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Can you make the option `unpackArray` case-insensitive?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: If we add this new option here, I feel we'd be better to document somewhere (e.g., `sq/functions.scala`)

**Comment by gatorsmile** [comment]
cc @gengliangwang

**Comment by gengliangwang** [comment]
retest this please.

**Comment by MaxGekk** [comment]
@maropu For now it is impossible to specify schema for `from_json` if it is not `StructType`. The PR https://github.com/apache/spark/pull/21472 solves the problem.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 2]
```diff
@@ -548,7 +553,9 @@ case class JsonToStructs(
       forceNullableSchema = SQLConf.get.getConf(SQLConf.FROM_JSON_FORCE_NULLABLE_SCHEMA))
 
   override def checkInputDataTypes(): TypeCheckResult = nullableSchema match {
-    case _: StructType | ArrayType(_: StructType, _) | _: MapType =>
+    case ArrayType(_: StructType, _) if unpackArray =>
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Even if `unpackArray` is `false`, the next branch in line 558 still do `super.checkInputDataTypes()` for any `ArrayType`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 3]
```diff
@@ -101,6 +102,13 @@ class JacksonParser(
     }
   }
 
+  private def makeArrayRootConverter(at: ArrayType): JsonParser => Seq[InternalRow] = {
+    val elemConverter = makeConverter(at.elementType)
+    (parser: JsonParser) => parseJsonToken[Seq[InternalRow]](parser, at) {
+      case START_ARRAY => Seq(InternalRow(convertArray(parser, elemConverter)))
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: In line 87:
```
        val array = convertArray(parser, elementConverter)
        // Here, as we support reading top level JSON arrays and take every element
        // in such an array as a row, this case is possible.
        if (array.numElements() == 0) {
          Nil
        } else {
          array.toArray[InternalRow](schema).toSeq
        }
```
Should we also follow this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 4]
```diff
@@ -523,6 +523,11 @@ case class JsonToStructs(
   // can generate incorrect files if values are missing in columns declared as non-nullable.
   val nullableSchema = if (forceNullableSchema) schema.asNullable else schema
 
+  private val caseInsensitiveOptions = CaseInsensitiveMap(options)
+  private val unpackArray: Boolean = {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Why do we need this? Can you add comments about it?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 5]
```diff
@@ -423,7 +423,9 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
     val input = """{"a": 1}"""
     val schema = ArrayType(StructType(StructField("a", IntegerType) :: Nil))
     val output = InternalRow(1) :: Nil
-    checkEvaluation(JsonToStructs(schema, Map.empty, Literal(input), gmtId, true), output)
+    checkEvaluation(
+      JsonToStructs(schema, Map("unpackArray" -> "true"), Literal(input), gmtId, true),
```
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: add case for `unpackArray` as `false`

**Comment by maropu** [comment]
> What kind of tests would you expect in json-functions.sql. Probably you would expect tests
> that are different from added to JsonExpressionsSuite.scala.
IIUC there is no strict rule there and we tend to put SQL related tests in `SQLQueryTestSuite`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: The code in line 87 returns `null` for json input `[]` if schema is `StructType(StructField("a", IntegerType) :: Nil)`. I would explain why we should return `null` in that case: we *extract* struct from the array. If the array is _empty_, it means there is nothing to extract and we returns `null` for the nothing.

In case when schema is `ArrayType(...)`,  I believe we should return `empty` array for empty JSON array `[]`

**Comment by MaxGekk** [comment]
@maropu I need changes from this PR https://github.com/apache/spark/pull/21550 (or this https://github.com/apache/spark/pull/21472) to write SQL tests for `ArrayType`

**Comment by MaxGekk** [comment]
@maropu @gengliangwang I added SQL tests. Please, take a look at the PR again, please.

**Comment by MaxGekk** [comment]
@maropu @gengliangwang May I ask you to look at the PR one more time.

**Comment by MaxGekk** [comment]
@gatorsmile May I ask you to look at the PR.

**Comment by MaxGekk** [comment]
@gatorsmile Could you tell me, please, what does prevent the PR from getting merged?

**Comment by gengliangwang** [comment]
I guess it is still controversial to have this new behavior with the new option.

**Comment by MaxGekk** [comment]
I set the option to the value which keep current behavior. So, it should be absolutely compatibly with current implementation

**Comment by MaxGekk** [comment]
@gatorsmile @gengliangwang @maropu The change doesn't break existing behavior. I set new option to the value which preserve backward compatibly. The PR just extend existing implementation and supports arrays of any types. May I ask you to look at the PR again.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 6]
```diff
@@ -61,6 +61,7 @@ class JacksonParser(
     dt match {
       case st: StructType => makeStructRootConverter(st)
       case mt: MapType => makeMapRootConverter(mt)
+      case at: ArrayType => makeArrayRootConverter(at)
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: > This change accepts the json datasource form that the master can't parse?

Right, it accept arrays of any types comparing to the master which accepts arrays of structs only

> If so, I think we need tests in JsonSuite ...

I added a few tests to `JsonSuite` or do you mean some concrete test case?

[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: You've already added tests in `JsonSuite`? It seems there are tests in `JsonFunctionsSuite`, `JsonExpressionSuite`, and `SQLQueryTestSuite` now?

[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: Sorry, I didn't catch you meant specific class. Just in case, what is the reason for adding tests to `JsonSuite`? I changed behavior of a function, so, `JsonFunctionsSuite` is perfect place for new tests, I believe. At the moment, you cannot specify a schema of json objects different from `StructType` in `DataFrameReader`. In this way, the `ArrayType` can come into `JacksonParser` only from json functions (`from_json`). I can add similar test to `JsonSuite` as in `JsonFunctionsSuite` but it doesn't make any sense from my point of view.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 7]
```diff
@@ -423,7 +423,9 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
     val input = """{"a": 1}"""
     val schema = ArrayType(StructType(StructField("a", IntegerType) :: Nil))
     val output = InternalRow(1) :: Nil
-    checkEvaluation(JsonToStructs(schema, Map.empty, Literal(input), gmtId), output)
+    checkEvaluation(
+      JsonToStructs(schema, Map.empty, Literal(input), gmtId),
+      output)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: Looks unrelated change.

**Comment by HyukjinKwon** [comment]
How does it relate to https://github.com/apache/spark/pull/19492?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: ah, ok. You touched the the `JacksonParser.scala` file, so I though there were some behaivour changes in the json datasorce. But I notice that there are not, so the current tests are enough. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 8]
```diff
@@ -101,6 +102,17 @@ class JacksonParser(
     }
   }
 
+  private def makeArrayRootConverter(at: ArrayType): JsonParser => Seq[InternalRow] = {
+    val elemConverter = makeConverter(at.elementType)
+    (parser: JsonParser) => parseJsonToken[Seq[InternalRow]](parser, at) {
+      case START_ARRAY => Seq(InternalRow(convertArray(parser, elemConverter)))
+      case START_OBJECT if at.elementType.isInstanceOf[StructType] =>
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: This is for backward compatibility to the behavior introduced by this PR: https://github.com/apache/spark/pull/16929,  most likely by this: https://github.com/apache/spark/pull/16929/files#diff-6626026091295ad8c0dfb66ecbcd04b1R506 . Even special test was added: https://github.com/apache/spark/pull/16929/files#diff-88230f171af0b7a40791a867f9dd3a36R382 . Please, ask author @HyukjinKwon about the it.

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: If an array of data is empty, and the schema is an array, it should return an empty array, https://github.com/apache/spark/pull/16929/files#diff-88230f171af0b7a40791a867f9dd3a36R389

[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: I think it is a super weird case...can we put this special handling code for back-compatibility in `JsonToStructs`?
Ether way, I think we should leave code comments here to make others understood easily?

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: I think what was done in that PR is pretty different and it is actually the opposite of what we are doing here. Indeed, there we are returning an array of structs when a struct is specified as schema and the JSON contains an array. Here we are returning an array with one struct when the schema is an array of struct and there is a struct instead of an array.

Despite I don't really like the behavior introduced in the PR you mentioned, I can understand it, as it was a way to support array of struct (the only at the moment) and I don't think we can change it before 3.0 at least for backward compatibility. But since here we are introducing a new behavior, if an array is required and a struct is found, I think returning an array with one element is a wrong/unexpected behavior and returning null would be what I'd expect as a user.

[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: I can only say that this `case START_OBJECT` was added to handle the case when an user specified a schema as an `array` of `struct`, and a `struct` is found in the input json. See the existing test case which I pointed out above: https://github.com/apache/spark/pull/16929/files#diff-88230f171af0b7a40791a867f9dd3a36R382 . I don't want to change the behavior in the PR and potentially break user's apps. What I would propose is to put the functionality under a `spark.sql.legacy.*` flag which could be deleted in Spark 3.0.

@maropu Initially I put the behavior under a flag in `JsonToStructs` but Reynold asked me to support new behavior without any new sticks/flags.

[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: Shall we add a comment on top of this `case` to explain it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 9]
```diff
@@ -544,34 +544,27 @@ case class JsonToStructs(
       timeZoneId = None)
 
   override def checkInputDataTypes(): TypeCheckResult = nullableSchema match {
-    case _: StructType | ArrayType(_: StructType, _) | _: MapType =>
+    case _: StructType | _: ArrayType | _: MapType =>
       super.checkInputDataTypes()
     case _ => TypeCheckResult.TypeCheckFailure(
       s"Input schema ${nullableSchema.catalogString} must be a struct or an array of structs.")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: `or an array of structs.` -> `or an array.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 10]
```diff
@@ -544,34 +544,27 @@ case class JsonToStructs(
       timeZoneId = None)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Pleas also update the comment of `JsonToStructs`:

`Converts an json input string to a [[StructType]] or [[ArrayType]] of [[StructType]]s`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: I am adding a comment for this.

**Comment by MaxGekk** [comment]
Is there anything for now which blocks the PR?

**Comment by MaxGekk** [comment]
@gatorsmile @HyukjinKwon May I ask you to look at the PR one more time.

**Comment by MaxGekk** [comment]
@HyukjinKwon Are there any chances the PR will be merged? or I should close it?

**Comment by MaxGekk** [comment]
@gatorsmile Could you look at the PR, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 11]
```diff
@@ -101,6 +102,21 @@ class JacksonParser(
...
+        // This handles the case when an input JSON object is a structure but
+        // the specified schema is an array of structures. In that case, the input JSON is
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: Could you add an example here, like what we did in `makeStructRootConverter `?

File: sql/core/src/test/resources/sql-tests/inputs/json-functions.sql [hunk 12]
```diff
@@ -39,3 +39,8 @@ select from_json('{"a":1, "b":"2"}', 'struct<a:int,b:string>');
 -- infer schema of json literal
 select schema_of_json('{"c1":0, "c2":[1]}');
 select from_json('{"c1":[1, 2, 3]}', schema_of_json('{"c1":[0]}'));
+
+-- from_json - array type
+select from_json('[1, 2, 3]', 'array<int>');
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/json-functions.sql]: Add more cases ?
select from_json('[3, null, 4]', 'array<int>')
select from_json('[3, "str", 4]', 'array<int>')

**Comment by gatorsmile** [comment]
LGTM

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
LGTM too.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by viirya** [comment]
I think R side is not update for this yet. @huaxingao would you like to do that?

**Comment by huaxingao** [comment]
Sure. I will work on it. Thanks for letting me know. @viirya
